src/spi2.rs

```diff
@@ -0,0 +1,681 @@
+//! SPI peripheral control
+//!
+//! Currently only implements full duplex controller mode support.
+//!
+//! SPI0 is reserved for accessing flash and sram and therefore not usable for other purposes.
+//! SPI1 shares its external pins with SPI0 and therefore has severe restrictions in use.
+//!
+//! SPI2 & 3 can be used freely.
+//!
+//! The CS pin is controlled by hardware on esp32 (contrary to the description of embedded_hal).
+//!
+//! The [Transfer::transfer], [Write::write] and [WriteIter::write_iter] functions lock the
+//! APB frequency and therefore the requests are always run at the requested baudrate.
+//! The primitive [FullDuplex::read] and [FullDuplex::send] do not lock the APB frequency and
+//! therefore may run at a different frequency.
+//!
+//! # TODO
+//! - Quad SPI
+//! - DMA
+//! - Multiple CS pins
+//! - Slave
+
+// TODO -> transfer size & DMA
+// update usage example
+// add update device config Fn
+// &mut self in transaction only needed for sw_cs
+
+use core::cmp::{max, min, Ordering};
+use core::marker::PhantomData;
+use core::ptr;
+
+use std::borrow::Borrow;
+use std::cell::RefCell;
+use std::sync::Mutex;
+
+use embedded_hal::spi::{SpiBus, SpiBusFlush, SpiBusRead, SpiBusWrite, SpiDevice};
+
+use esp_idf_sys::*;
+
+use crate::delay::BLOCK;
+use crate::gpio::{self, InputPin, OutputPin, Output,InputOutput, Pin,AnyIOPin, AnyOutputPin, IOPin, PinDriver};
```

Comment:
Formatting.

Reply:
yeah, my code is still ugly :see_no_evil: it was not 100 % ready for a review